### PR TITLE
Memory trial

### PR DESCRIPTION
**Pull request by yramachers**
So, this has a few changes with suspects leaking memory hopefully repaired. The biggest effect is to filter the output data in the same way the root script does. This should result in all the data required from the simulation but that's for a later stage. This is the code commit level that produced the valgrind output.

### CODE REVIEW

**Comment by yramachers** [comment]
Don't know about the memory use per event but the observation is that switching off biasing (factor default 1) not only speeds up the simulation as expected but also reduces the memory footprint significantly, maybe only because there is hardly time to watch it grow - to be tested. Biasing of course multiplies the number of tracks per event so memory use should be bigger.

**Comment by yramachers** [comment]
More test runs: Biasing neutron production has the biggest effect on runtime and memory usage, as expected. For a factor ten increase in neutron prod sigma from 100 to 1000 bias memory max stabilises at just under a factor 2 increase, from about 25% to 47% on my SCRTP machine. Runtime rises by roughly a factor 8 while the hit number modulo statistics is consistent with an increase by a factor 10. That's not too bad. It also means that brute force, no bias, runs will need roughly 10^6 events to see any appreciable number of hits. Maybe these can be gradually lifted by only increasing the neutron capture sigma on Ge which does not seem to make much difference for runtime or memory (it is hardly ever needed, no wonder). 
I changed the root script completely since the mass info output is really not feasible. There are far too many trajectories with no value in that set hence filtering before output is the correct way to go.

**Comment by drbenmorgan** [comment]
I've had a few run throughs on the current `master` branch, and I think I've traced the basic issue down to the trajectory storage. I don't *think* there's a direct memory leak here, more that the number of trajectories (and hence trajectory points) gets very large. As these are memory managed through the `G4Allocator` mechanism, I can see these get up to quite large buffer sizes. 

I can submit a separate PR with a workaround in, but this means optionally switching of trajectory storage. There may well be a true leak here somewhere, but it's definitely with the Trajectories.

**Comment by yramachers** [comment]
This rebasing is quite confusing at first. Easy to use but then it shows conflicts and refuses to rebase, so I ... skipped. That went fine at the end but all the filtering of event data output was gone. Not a smart move from my side. Fortunately, github was still as before and resolving the merge conflict here is soooo much easier. Resolved conflict, committed as requested and then pulled from the branch on github and I was back on my git repository where I thought I'd get with rebase. Should be fine. Did you notice that the test without trajectory storage runs a factor 20 faster? Anyway, all passes on SCRTP.

File: include/WLGDEventAction.hh [hunk 1]
```diff
@@ -47,6 +45,27 @@ private:
   G4THitsMap<G4double>*      GetHitsCollection(G4int hcID, const G4Event* event) const;
   G4THitsMap<G4ThreeVector>* GetVecHitsCollection(G4int hcID, const G4Event* event) const;
 
+  std::vector<int> FilterTrajectories(int item, std::vector<G4int>& tid, std::vector<G4int>& pid)
```
Comment (drbenmorgan) [inline include/WLGDEventAction.hh:48]: I think a brief outline of what this function does, including the inputs/outputs, would be worthwhile. For future-proofing, the doxygen style can be used:

```c++
//! Brief description
/*!
 * More detail if needed
 *
 * \param[in] item Description of parameter
 * \param[in] tid Description ...
 * \param[in] pid Description ... 
 */
std::vector<int> FilterTrajectories(int item, std::vector<G4int>& tid, std::vector<G4int>& pid)
```

I *think* the `tid` and `pid` parameters can be `const`

File: include/WLGDEventAction.hh [hunk 2]
```diff
@@ -47,6 +45,27 @@ private:
...
+  {
+    int idx = 0;
+    int pidx = 0;
+    std::vector<int> result;
+    std::vector<int>::iterator it;
+
+    it = std::find(tid.begin(), tid.end(), item);
```
Comment (drbenmorgan) [inline include/WLGDEventAction.hh:55]:
```suggestion
    auto it = std::find(tid.begin(), tid.end(), item);
```

File: include/WLGDEventAction.hh (hunk 1 above)
Comment (yramachers) [reply include/WLGDEventAction.hh]: Latest commit adds doxygen description of filter function.

**Pull request merged**